### PR TITLE
added a basic implementation of bst

### PR DESCRIPTION
**Pull request by godarkthrt**

### CODE REVIEW

File: day1/bst.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+func addRecursive(curr *node, target int) (*node, bool) {
+	var ok bool
+	if curr == nil {
+		curr, ok = &node{val: target}, true
+	} else if target < curr.val {
+		curr.left, ok = addRecursive(curr.left, target)
+	} else if target > curr.val {
+		curr.right, ok = addRecursive(curr.right, target)
+	}
```
Comment (kortschak) [inline day1/bst.go:50]: Again, I'd use returns, though now not for all.

```
if curr == nil {
    return &node{val: target}, true
}
switch {
case target < curr.val:
    curr.left, ok = addRecursive(curr.left, target)
case target > curr.val:
    curr.right, ok = addRecursive(curr.right, target)
}
return curr, ok
```

File: day1/bst.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+func (b *Bst) Remove(target int) bool {
+	var ok bool
+	if b == nil {
+		return false
+	} else {
+		b.root, ok = removeRecursive(b.root, target)
+		if ok {
+			b.size--
+		}
+	}
```
Comment (kortschak) [inline day1/bst.go:64]: Remove else and dedent.

File: day1/bst.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			curr, ok = curr.left, true
+		} else {
+			// we need to find successor node , switch with it and then delete it
+			s := findSuccessor(curr.right)
+			// swap the data and call easy recursive remove
+			curr.val, s.val = s.val, curr.val
+			curr.right, ok = removeRecursive(curr.right, target)
+
+		}
+	}
```
Comment (kortschak) [inline day1/bst.go:92]:
```
if curr == nil {
	return curr, false
}
switch {
case target < curr.val:
	curr.left, ok = removeRecursive(curr.left, target)
case if target > curr.val:
	curr.right, ok = removeRecursive(curr.right, target)
default:
	// this is the case of actual delete
	if curr.left == nil && curr.right == nil {
		return nil, true
	}
	if curr.left == nil {
		return curr.right, true
	}
	if curr.right == nil {
		return curr.left, true
	}

	// we need to find successor node , switch with it and then delete it
	s := findSuccessor(curr.right)
	// swap the data and call easy recursive remove
	curr.val, s.val = s.val, curr.val
	curr.right, ok = removeRecursive(curr.right, target)
}
```

File: day1/bst.go (hunk 1 above)
Comment (godarkthrt) [reply day1/bst.go]: yeah makes sense this is making code for cleaner

**Pull request merged**